pyomo/contrib/pynumero/sparse/mpi_block_matrix.py

```diff
@@ -764,38 +764,145 @@ def __sub__(self, other):
     def __rsub__(self, other):
         raise NotImplementedError('Operation not supported by MPIBlockMatrix')
 
-    def _block_vector_multiply(self, other):
-        """
-        Parameters
-        ----------
-        other: BlockVector
-
-        Returns
-        -------
-        result: BlockVector
-        """
-        block_vector_assert_block_structure(other)
-        assert self.bshape[1] == other.nblocks, 'Dimension mismatch'
-        local_result = BlockVector(self.bshape[0])
-        for row_ndx in range(self.bshape[0]):
-            local_result.set_block(row_ndx, np.zeros(self.get_row_size(row_ndx)))
-        rank = self._mpiw.Get_rank()
+    def _get_block_vector_for_dot_product(self, x):
+        if isinstance(x, MPIBlockVector):
+            """
+            Consider a non-empty block m_{i, j} from the mpi block matrix with rank owner r_m and the 
+            corresponding block v_{j} from the mpi block vector with rank owner r_v. There are 4 cases:
+              1. r_m = r_v
+                 In this case, all is good.
+              2. r_v = -1
+                 In this case, all is good.
+              3. r_m = -1 and r_v = 0
+                 All is good
+              4. If none of the above cases hold, then v_{j} must be broadcast
+            """
+            n_block_rows, n_block_cols = self.bshape
+            blocks_needing_broadcast = np.zeros(n_block_cols, dtype=np.int64)  # a value > 0 means broadcast
+            x_rank_ownership = x.rank_ownership
+            comm = self._mpiw
+            rank = comm.Get_rank()
+
+            if rank == 0:
+                block_indices = self._owned_mask
+            else:
+                block_indices = self._unique_owned_mask
+            block_indices = np.bitwise_and(block_indices, self._block_matrix._block_mask)
+            for i, j in zip(*np.nonzero(block_indices)):
+                r_m = self._rank_owner[i, j]
+                r_v = x_rank_ownership[j]
+                if r_m == r_v:
+                    pass
+                elif r_v == -1:
+                    pass
+                elif r_m == -1 and r_v == 0:
+                    pass
```

Comment:
It would be nice to use named token (e.g. `ALL_OWNED`) instead of embedded constants here

Reply:
I agree. This really needs changed all over `mpi_block_vectory.py` and `mpi_block_matrix.py`